content/blog/2022-02-22-profiling.md

```diff
@@ -0,0 +1,208 @@
+
+
+Table 1 shows the experiment results on runtime overhead of path profiling (PP) and edge profiling (QPT). The runtime overhead of path profiling is 30.9%, while the overhead of edge profiling is 16.1%. 
+
+
+#### Is the overhead acceptable?
+
+We discussed it with the conclusion that it depends (oh, the ultimate answer!). 
+
+It depends on the application and how frequently you need to use path profiling. For example, if we use it in JIT compilation and only profile it once, then 30% overhead is acceptable for its improving performance. 
+
+It’s also worth noting that the result presented in this paper might be the worst-case scenario of runtime overhead because the compiled SPEC benchmarks execute very fast. But in real cases, we may profile at the bytecode level. Then the overhead ratio might be lower because bytecode execution is slower. 
+
+
+### Limitations
+
+
+#### Is it efficient enough?
+
+We need to run edge profiling first and determine where we will put our instrumentation. Only after that, can we run path profiling. This additional overhead makes readers think that the Ball-Larus method would not be as efficient as it claims in the paper. 
+
+Some other works also raise questions on it: Kumar et al. [1] claim that methods like Ball-Larus algorithm take significant time to identify profile points in the program, cannot be used in dynamic (JIT) compilation. 
+
+We think the phase of identifying instrumentation could be accelerated by only analyzing a tiny and representative sample of the entire program because we only need a rough idea for the edge profiling result. There is also an interesting idea to adopt static program analysis methods like “constant propagation” to know some deterministic value range of certain variables to know additional information about edge profiling without any runtime analysis. 
+
+
+### Work built upon this method
+
+
+#### What applications will benefit from accurate Path Profiling?
+
+ Profile-driven compilers and software test coverage are two applications that would benefit from accurate path profiling mentioned in the paper. We further discussed how to perform the optimizations given the path profiling information. 
+
+
+
+1. Profile-driven compilers
+
+    JIT (Just-In-Time) compilation: understand the hot path and do optimization like superblock (straight-line code with side exits).
+
+2. Software test coverage
+
+    Quantify the adequacy of a test dataset by profiling a program and reporting unexecuted statements or control-flow. 
+
+
+
+#### Research work
+
+For functions with many paths, allocating an array entry for all program paths is prohibitively expensive, if not feasible [2]. Vaswani et al. [1] propose preferential path profiling (PPP): separating interesting paths from other paths and assigning a set of unique and compact numbers to these interesting paths. This enables the PPP implementation to record path information in an array instead of a hash table. 
```

Comment:
Since this is a blog post, maybe just link to the papers instead of (or in addition to?) using numbered citations like `[1]`.

Reply:
I changed to using both, clicking on the numbered citations will jump to the papers